Http::get - use noProxy for external requests

macbre opened the pull request:
[PLATFORM-1856](https://wikia-inc.atlassian.net/browse/PLATFORM-1856)

As we're planning to get rid of local Varnish instance running on Apache nodes, we want to use  border nodes as HTTP proxies. However, we do not want to use them when making external HTTP requests - hence this change

Introduce `ExternalHttp` helper class with `get` and `post` methods that will set `noProxy` option to true internally and clearly state what kind of HTTP request we want to perform here. In the near future we'll add more headers for tracing internal HTTP requests.

@wladekb / @artursitarski

macbre committed:
Http::get - use noProxy for external requests

macbre commented on extensions/wikia/Recirculation/services/FandomDataService.class.php:
```diff
@@ -55,7 +55,7 @@ private function apiRequest( $type ) {
 
 		$url = $this->buildUrl( $endpoint, $options );
-		$data = Http::get( $url );
+		$data = Http::get( $url, null, [ 'noProxy' => true ] );
```
Pinging @pauloslund

wladekb commented on extensions/3rdparty/LyricWiki/LyricFind/LyricFindTrackingService.class.php:
```diff
@@ -122,7 +122,10 @@ public function track($amgId, $gracenoteId, Title $title) {
 		wfDebug(__METHOD__ . ': ' . json_encode($data) . "\n");
 
-		$resp = Http::post($url, ['postData' => $data]);
+		$resp = Http::post($url, [
```
I wonder about creating something like `ExternalHttp::post` which would also help us when we get to implementing HTTP headers for request tracing, what do you think? or maybe leave `Http` as it is and introduce `InternalHttp` ?

macbre replied:
We perform more internal requests than external ones (by both number of requests and places in the code that make them). I was thinking about introducing `ExternalHttp` helper class with `get` and `post` methods. With the "HTTP headers for request tracing" we have a second use case for this helper class. Let's implement it.

macbre replied:
Introduced in 6b045ed

macbre committed:
Introduce ExternalHttp

Use ExternalHttp to make external HTTP requests. This will by-pass the Wikia HTTP proxy and skip adding our custom HTTP headers used for tracing requests.

See PLATFORM-1856

macbre committed:
Use ExternalHttp instead of Http::get + noProxy combo

macbre commented:
```
LyricFindTrackingServiceTest::testTrackResponseCode with data set #4 (1234, array(array(206)), true)
API response code should match expected value
Failed asserting that false matches expected true.
```

wladekb commented:
lgtm

macbre committed:
LyricFindTrackingServiceTest: fix mocking of Http request response

macbre committed:
AdEngine: use ExternalHttp::get

macbre committed:
Use ExternalHttp instead of Http::get + noProxy

macbre commented on extensions/wikia/AdEngine/ResourceLoaders/ResourceLoaderAdEngineBase.php:
```diff
@@ -23,7 +23,7 @@
 	 */
 	protected function fetchRemoteScript( $url ) {
-		return Http::get( $url, 'default', [ 'noProxy' => true ] );
+		return ExternalHttp::get( $url );
```
@nandy-andy - please review

macbre commented on extensions/wikia/AdEngine/ResourceLoaders/ResourceLoaderAdEngineSourcePointRecoveryModule.php:
```diff
@@ -41,11 +41,10 @@ protected function getFallbackDataWhenRequestFails() {
 		global $wgSourcePointApiKey;
 
-		$content = Http::get( $url,
+		$content = ExternalHttp::get( $url,
```
@nandy-andy - please review

macbre commented on extensions/wikia/Recirculation/services/FandomDataService.class.php:
```diff
@@ -55,7 +55,7 @@ private function apiRequest( $type ) {
 
 		$url = $this->buildUrl( $endpoint, $options );
-		$data = Http::get( $url );
+		$data = ExternalHttp::get( $url );
```
@pauloslund

pauloslund replied:
:+1:

nandy-andy commented:
It looks good. However, I'm surprised there is no files from `WikiaMaps` extension. I'll double-check tomorrow and get back to it.

macbre commented:
I did not replace calls to `Http::get` with `noProxy` option set that DID not perform external requests. This was the case mostly for internal services calls (to bypass the local Varnish instance causing problems).

nandy-andy commented:
Oh, ok. Thanks for update! :) Therefore, it looks good to me :+1:

gabrys commented:
Why not doing this the other way round. I assume connecting to an internal HTTP service is a special case, not connecting to an external one

macbre commented:
@rychu, external traffic is actually something quite extraordinary in our case.
